File: 백준/Gold/13460. 구슬 탈출 2/구슬 탈출 2.java
import java.util.*;

public class Main {
    static int n, m;
    static char[][] board;
    static int[] blank; // 구멍 위치
    static int[][][][] dist; // 방문 및 거리 기록

    static int[] dr = {0, 1, 0, -1}; // 오른쪽, 아래, 왼쪽, 위
    static int[] dc = {1, 0, -1, 0};

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        m = sc.nextInt();
        board = new char[n][m];

        int[] red = new int[2];
        int[] blue = new int[2];
        blank = new int[2];

        for (int i = 0; i < n; i++) {
            String line = sc.next();
            for (int j = 0; j < m; j++) {
                board[i][j] = line.charAt(j);
                if (board[i][j] == 'R') {
                    red[0] = i;
                    red[1] = j;
                    board[i][j] = '.';
                }
                if (board[i][j] == 'B') {
                    blue[0] = i;
                    blue[1] = j;
                    board[i][j] = '.';
                }
                if (board[i][j] == 'O') {
                    blank[0] = i;
                    blank[1] = j;
                }
            }
        }

        System.out.println(bfs(red[0], red[1], blue[0], blue[1]));
    }

    static int bfs(int rr, int rc, int br, int bc) {
        dist = new int[n][m][n][m];
        for (int[][][] a : dist) {
            for (int[][] b : a) {
                for (int[] c : b) {
                    Arrays.fill(c, Integer.MAX_VALUE);
                }
            }
        }

        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{rr, rc, br, bc});
        dist[rr][rc][br][bc] = 0;

        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int r1 = cur[0], c1 = cur[1], r2 = cur[2], c2 = cur[3];
            int curDist = dist[r1][c1][r2][c2];

            if (curDist > 10) return -1;
            if (r1 == blank[0] && c1 == blank[1]) return curDist;

            for (int d = 0; d < 4; d++) {
                int[] next = moveBalls(r1, c1, r2, c2, d);
                int nr1 = next[0], nc1 = next[1], nr2 = next[2], nc2 = next[3];

                if (nr2 == blank[0] && nc2 == blank[1]) continue; // 파란 구슬이 구멍에 빠짐
                if (nr1 == -1) continue; // 둘 다 동시에 구멍에 빠짐

                if (dist[nr1][nc1][nr2][nc2] == Integer.MAX_VALUE) {
                    dist[nr1][nc1][nr2][nc2] = curDist + 1;
                    q.add(new int[]{nr1, nc1, nr2, nc2});
                }
            }
        }
        return -1;
    }

    static int[] moveBalls(int rr, int rc, int br, int bc, int dir) {
        int[] redMoved = move(rr, rc, dir);
        int[] blueMoved = move(br, bc, dir);

        int nrr = redMoved[0], nrc = redMoved[1];
        int nbr = blueMoved[0], nbc = blueMoved[1];

        // 둘 다 동시에 구멍에 빠짐
        if (nrr == blank[0] && nrc == blank[1] && nbr == blank[0] && nbc == blank[1]) {
            return new int[]{-1, -1, -1, -1};
        }

        // 같은 칸에 겹침 (구멍 제외)
        if (nrr == nbr && nrc == nbc) {
            if (nrr == blank[0] && nrc == blank[1]) {
                return new int[]{-1, -1, -1, -1};
            }
            // 이동 순서에 따라 보정
            if (dir == 0) { // 오른쪽
                if (rc > bc) nbc -= 1;
                else nrc -= 1;
            } else if (dir == 1) { // 아래
                if (rr > br) nbr -= 1;
                else nrr -= 1;
            } else if (dir == 2) { // 왼쪽
                if (rc < bc) nbc += 1;
                else nrc += 1;
            } else { // 위
                if (rr < br) nbr += 1;
                else nrr += 1;
            }
        }

        return new int[]{nrr, nrc, nbr, nbc};
    }

    static int[] move(int r, int c, int dir) {
        while (true) {
            int nr = r + dr[dir];
            int nc = c + dc[dir];
            if (board[nr][nc] == '#') break;
            r = nr;
            c = nc;
            if (r == blank[0] && c == blank[1]) break;
        }
        return new int[]{r, c};
    }
}

File: 백준/Gold/13460. 구슬 탈출 2/구슬 탈출 2.py
n, m = map(int, input().split())
board = [list(input()) for _ in range(n)]

from collections import deque

def solution(n, m, board):
    
    def get_init(x):
        for i in range(n):
            for j in range(m):
                if x == board[i][j]:
                    return (i, j)
    
    def move_right(r, c):
        for i in range(c, m):
            if board[r][i] == '.':
                c = i
                continue
            if board[r][i] == '#':
                c = i -1
                break
            if board[r][i] == 'O':
                c = i
                break
        return r, c

    def move_left(r, c):
        for i in reversed(range(c)):
            if board[r][i] == '.':
                c = i
                continue
            if board[r][i] == '#':
                c = i + 1
                break
            if board[r][i] == 'O':
                c = i
                break
        return r, c

    def move_up(r, c):
        for i in reversed(range(r)):
            if board[i][c] == '.':
                r = i
                continue
            if board[i][c] == '#':
                r = i + 1
                break
            if board[i][c] == 'O':
                r = i
                break
        return r, c

    def move_down(r, c):
        for i in range(r, n):
            if board[i][c] == '.':
                r = i
                continue
            if board[i][c] == '#':
                r = i - 1
                break
            if board[i][c] == 'O':
                r = i
                break
        return r, c

    
    def go_right(rr, rc, br, bc):
        nrr, nrc = move_right(rr, rc)
        nbr, nbc = move_right(br, bc)
        if (nrr, nrc) == (nbr, nbc): # 공이 같은 자리에 있을 때
            if (nrr, nrc) == blank: # 공이 동시에 구멍에 빠진 상황: 금지된 상황
                return (-1, -1, -1, -1)
            # 공이 겹친상황: 보정이 필요
            if rc < bc:
                nrc -= 1
            elif rc > bc:
                nbc -= 1
        return nrr, nrc, nbr, nbc
    
    def go_left(rr, rc, br, bc):
        nrr, nrc = move_left(rr, rc)
        nbr, nbc = move_left(br, bc)
        if (nrr, nrc) == (nbr, nbc): # 공이 같은 자리에 있을 때
            if (nrr, nrc) == blank: # 공이 동시에 구멍에 빠진 상황: 금지된 상황
                return (-1, -1, -1, -1)
            # 공이 겹친상황: 보정이 필요
            if rc < bc: # R 1, B 4 -> R 1, B 1 -> R 1, B 2
                nbc += 1
            elif rc > bc:
                nrc += 1
        return nrr, nrc, nbr, nbc
    
    def go_up(rr, rc, br, bc):
        nrr, nrc = move_up(rr, rc)
        nbr, nbc = move_up(br, bc)
        if (nrr, nrc) == (nbr, nbc): # 공이 같은 자리에 있을 때
            if (nrr, nrc) == blank: # 공이 동시에 구멍에 빠진 상황: 금지된 상황
                return (-1, -1, -1, -1)
            # 공이 겹친상황: 보정이 필요
            if rr < br:
                nbr += 1
            elif rr > br:
                nrr += 1
        return nrr, nrc, nbr, nbc
    
    def go_down(rr, rc, br, bc):
        nrr, nrc = move_down(rr, rc)
        nbr, nbc = move_down(br, bc)
        if (nrr, nrc) == (nbr, nbc): # 공이 같은 자리에 있을 때
            if (nrr, nrc) == blank: # 공이 동시에 구멍에 빠진 상황: 금지된 상황
                return (-1, -1, -1, -1)
            # 공이 겹친상황: 보정이 필요
            if rr < br:
                nrr -= 1
            elif rr > br:
                nbr -= 1
        return nrr, nrc, nbr, nbc
    
    def process(rr, rc, br, bc, dir):
        if dir == 0:
            return go_right(rr, rc, br, bc)
        if dir == 1:
            return go_down(rr, rc, br, bc)
        if dir == 2:
            return go_left(rr, rc, br, bc)
        if dir == 3:
            return go_up(rr, rc, br, bc)
        
    
    red, blue, blank = get_init('R'), get_init('B'), get_init('O')
    
    board[red[0]][red[1]] = '.'
    board[blue[0]][blue[1]] = '.'
    
    def bfs():
        
        dr = [-1, 1, 0, 0]
        dc = [0, 0, -1, 1]
        
        def in_range(r, c):
            return 0 <= r < n and 0 <= c < m 
        
        
        # (m -> n) -> m -> n
        # c2    r2    c1   r1
        in_queue = [
                    [
                    [[False for _ in range(m)] for _ in range(n)]
                    for _ in range(m)
                    ]
                    for _ in range(n)
                    ]
        inf = int(1e18)
        # 초기상태 (r1, c1) (r2, c2) -> 다음 상태 (nr1, nc1) (nr2, nc2) 까지의 최단거리 = dist[nr1][nc1][nr2][nc2]
        dist = [
                [
                [[inf for _ in range(m)] for _ in range(n)]
                for _ in range(m)
                ]
                for _ in range(n)
                ]
        
        dq = deque([])
        r1, c1, r2, c2 = red[0], red[1], blue[0], blue[1]
        dq.append((r1, c1, r2, c2))
        in_queue[r1][c1][r2][c2] = True
        dist[r1][c1][r2][c2] = 0
        
        while dq:
            rr, rc, br, bc = dq.popleft()
            
            # print(f"d: {d}, rr: {rr}, rc: {rc}, br: {br}, bc: {bc}")
            
            cur_dist = dist[rr][rc][br][bc]
            
            if cur_dist > 10:
                return -1
            
            if (rr, rc) == blank:
                return dist[rr][rc][br][bc]
            
            for k in range(4):
                nrr, nrc, nbr, nbc = process(rr, rc, br, bc, k)
                # print(f"process({rr}, {rc}, {br}, {bc}) = ", k, nrr, nrc, nbr, nbc)
                if (nbr, nbc) == blank:
                    continue
                if in_range(nrr, nrc) and in_range(nbr, nbc) and not in_queue[nrr][nrc][nbr][nbc]:
                    dq.append((nrr, nrc, nbr, nbc))
                    in_queue[nrr][nrc][nbr][nbc] = True
                    dist[nrr][nrc][nbr][nbc] = cur_dist + 1
            
        return -1

    res = bfs()
    
    return res

    
    
print(solution(n, m, board))